aten/src/ATen/native/CPUFallback.cpp

```diff
@@ -0,0 +1,141 @@
+#include <ATen/native/CPUFallback.h>
+
+#include <ATen/core/ivalue.h>
+#include <ATen/core/stack.h>
+#include <ATen/core/boxing/KernelFunction.h>
+#include <ATen/core/dispatch/Dispatcher.h>
+#include <torch/library.h>
+#include <ATen/Functions.h>
+#include <torch/library.h>
+
+namespace at { namespace native {
+
+// convenience helper for converting tensors to cpu
+
+std::vector<at::Tensor> to_cpu(const at::TensorList& tensors) {
+    // We can't just call at::to_cpu() on the entire list of Tensors
+    // Because it will break on undefined tensors. Separate out undefined tensors first.
+    std::vector<at::Tensor> cpu_tensors(tensors.size());
+    std::vector<at::Tensor> valid_tensors;
+    std::vector<bool> to_translate(tensors.size());
+    for (size_t i = 0; i < tensors.size(); ++i) {
+        const at::Tensor& tensor = tensors[i];
+        if (tensor.defined()) {
+            to_translate[i] = true;
+            valid_tensors.push_back(tensor);
+        } else {
+            cpu_tensors[i] = tensor;
+        }
+    }
+    auto cpu_valid_tensors = at::_to_cpu(valid_tensors);
+    for (size_t i = 0, defined_pos = 0; i < tensors.size(); ++i) {
+        if (to_translate[i]) {
+            cpu_tensors[i] = std::move(cpu_valid_tensors[defined_pos++]);
+        }
+    }
+  return cpu_tensors;
+}
+
+
+void cpu_fallback(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
+  auto& schema_args = op.schema().arguments();
+  const auto num_arguments = schema_args.size();
+  auto arguments = torch::jit::last(stack, num_arguments);
+  const auto arguments_begin = stack->size() - num_arguments;
+
+  std::vector<at::Tensor> tensor_args;
+  std::vector<int> tensor_args_indices;
+
+  // Step 1: Convert all non-CPU tensor inputs into CPU tensors
+  // and put them on the stack at the correct indices.
+  for (int64_t idx = 0; idx < arguments.size(); ++idx) {
+    const auto& ivalue = arguments[idx];
+    if (ivalue.isTensor()) {
+      tensor_args.push_back(ivalue.toTensor());
+      tensor_args_indices.push_back(idx);
+    } else if (ivalue.isTensorList()) {
+      // I'm worried about the ivalue storing a reference to the std::vector temporary that I create.
+      // To be extra safe I'm converting to a c10::List, IValue will take ownership of the items in the list.
+      auto cpu_ivalue = c10::IValue(c10::List<at::Tensor>(to_cpu(ivalue.toTensorList().vec())));
+      (*stack)[arguments_begin + idx] = std::move(cpu_ivalue);
+    }
+  }
+  auto cpu_tensors = to_cpu(tensor_args);
```

Comment:
It's a little wonky here that you don't arrange for all tensors to be converted to CPU at once; e.g., if you have `f(Tensor, TensorList)`, first TensorList will get materialized, then Tensor.

Reply:
> yeah I'd go further ask why we're collecting the individual tensor args into a temporary list here and passing it to the helper that has the extra list-related logic

This is actually needed for XLA. The individual xla tensor arguments can't be independently copied to CPU, there's some shared context between them involved with the copying. If you look at the xla implementation of `at::_to_cpu()`, it eventually [calls some fusing logic here](https://github.com/pytorch/xla/blob/280348439b23fd7aef5c57c9b92298f8890bcc98/torch_xla/csrc/tensor.cpp#L955).

> It's a little wonky here that you don't arrange for all tensors to be converted to CPU at once

Is the part you don't like the fact that arguments won't necessarily be copied to CPU in the order that they come in the operator's schema? That seems fair. But just to confirm, you're not suggesting a single `at::_to_cpu()` call with all of the Tensor + TensorList args?

Reply:
Merging all tensor args into one giant list is actually the best for XLA since in case some tensors share a part of underlying graph that part can be materialized only once. ;) We probably didn't care too much about perf here so we indeed did it separately in the current codegen, but I think it can be changed. 
In this PR I think it's fine if we document this inefficiency in the comment, also note that "materializing together" is a requirement specific to xla only". ;)

Reply:
I'm ok with adding a note: If the only factor is perf, there's also technically a tradeoff where eager backends that use the CPU fallback would get a small perf hit from the extra indexing logic to merge and slice up the list of tensors. But that's probably pretty mild, and I'm thinking of this as a knob we can use later to improve CPU fallback perf for XLA if we need to.